### PR TITLE
Add Multi-Grained Attention Network(MGAN) model

### PR DESCRIPTION
**Pull request by GeneZC**
Ignored the aspect alignment loss mentioned in the paper in order to maintain the structure of train.py

### CODE REVIEW

File: models/mgan.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        asp_pool = torch.sum(asp_out, dim=1)
+        asp_pool = torch.div(asp_pool, asp_len.float().unsqueeze(-1)).unsqueeze(-1) # batch_size x 2*hidden_dim x 1
+
+        alignment_mat = torch.matmul(ctx_out, asp_out.transpose(1, 2)) # batch_size x (ctx)seq_len x (asp)seq_len
+        # batch_size x 2*hidden_dim
+        f_asp2ctx = torch.matmul(ctx_out.transpose(1, 2), F.softmax(alignment_mat.max(2, keepdim=True)[0], dim=1)).squeeze(-1)
+        f_ctx2asp = torch.matmul(F.softmax(alignment_mat.max(1, keepdim=True)[0], dim=2), asp_out).transpose(1, 2).squeeze(-1) 
+
+        c_asp2ctx = torch.matmul(ctx_out.transpose(1, 2), F.softmax(torch.matmul(ctx_out, asp_pool), dim=1)).squeeze(-1) 
+        c_ctx2asp = torch.matmul(asp_out.transpose(1, 2), F.softmax(torch.matmul(asp_out, ctx_pool), dim=1)).squeeze(-1)
```
Comment (songyouwei) [inline models/mgan.py]: Compared with the original paper, attention weight matrix W_{ca} and W_{cc} is missing?

File: models/mgan.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        ctx_out, (_, _) = self.ctx_lstm(ctx, ctx_len) 
+        ctx_out = self.location(ctx_out, aspect_in_text) # batch_size x (ctx)seq_len x 2*hidden_dim
+        ctx_pool = torch.sum(ctx_out, dim=1)
+        ctx_pool = torch.div(ctx_pool, ctx_len.float().unsqueeze(-1)).unsqueeze(-1) # batch_size x 2*hidden_dim x 1
+
+        asp_out, (_, _) = self.asp_lstm(asp, asp_len) # batch_size x (asp)seq_len x 2*hidden_dim
+        asp_pool = torch.sum(asp_out, dim=1)
+        asp_pool = torch.div(asp_pool, asp_len.float().unsqueeze(-1)).unsqueeze(-1) # batch_size x 2*hidden_dim x 1
+
+        alignment_mat = torch.matmul(ctx_out, asp_out.transpose(1, 2)) # batch_size x (ctx)seq_len x (asp)seq_len
```
Comment (songyouwei) [inline models/mgan.py]: Inconsistent with the original paper

Comment (GeneZC) [reply models/mgan.py]: Yes, my implementation of attention is different from the original paper and I will fix it.
Thanks for reviewing.

Comment (GeneZC) [reply models/mgan.py]: BTW, your implementation of attention in IAN is also different from the paper as you've calculated attention weight in a method similar to the way mentioned in TRANSFORMER.

File: models/mgan.py (hunk 1 above)
Comment (GeneZC) [reply models/mgan.py]: Yes, the original paper calculate attention weight in bi-linear way while I'm using dot product.
Would it be better if I fix them?

Comment (GeneZC) [reply models/mgan.py]: Yes, the paper calculate attention weight in bi-linear way while I'm using dot product.
Would it be better if I fix them?

File: models/mgan.py (hunk 2 above)
Comment (songyouwei) [reply models/mgan.py]: The differences between these implementations are small and can be merged first.

I plan to add some parameters to the unified attention layer to increase the flexibility of use.
